### ISSUE TITLE
Handle ide, debug server, and no-debug disconnect

### ISSUE THREAD

**Issue by karthiknadig** [microsoft/ptvsd#1695]

**Comment by int19h** [comment]
At this point, we have three ways the debug session can end, with some further variations below:

1. IDE ends it:
    - by sending "disconnect" or "terminate";
    - by closing the message channel.
1. Debug server ends it:
    - by reporting "exited";
    - by closing the message channel.
1. Debuggee process ends it, by exiting.

Some of these are not entirely independent. For example, in the case where there is a debug server connection, the debuggee process exiting will cause the message channel to close, and so this can be handled under that item. But in any no-debug case, and also in the regular debug case where the debuggee process dies before we get a chance to establish a channel to the debug server inside of it, the last item is its own thing.

There's code in adapter to handle all the various situations covered the first two items, and partially the third one. However, this code is rather disorganized, and hard to follow, because it is split across several different handlers for the corresponding events. It is also completely unable to handle the no-debug case. 

This needs to be reorganized such that it is all in a single place that orchestrates the proper termination sequence between all the components that are still connected. This approach would also then fix no-debug.
